### PR TITLE
Feature/arestor

### PR DESCRIPTION
**Pull request by mateimicu**

### CODE REVIEW

File: argus/config/arestor.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2016 Cloudbase Solutions Srl
```
Comment (ader1990) [inline argus/config/arestor.py]: 2017

File: argus/recipes/cloud/windows.py [hunk 2]
```diff
@@ -260,6 +270,51 @@ def wait_cbinit_finalization(self):
         LOG.debug("Wait for the Cloudbase-Init service to stop ...")
         self._backend.remote_client.manager.wait_cbinit_service()
 
+    @staticmethod
+    def _get_namespace(service_type):
+        """Return the metadata namespace."""
+        # HACK(mmicu): for Openstack we have the service_type set to http
```
Comment (ader1990) [inline argus/recipes/cloud/windows.py]: Put a note here, no need for hack

File: argus/util.py [hunk 3]
```diff
@@ -384,6 +386,17 @@ def get_command(command, command_type=None):
 }
 
 
+@contextlib.contextmanager
```
Comment (ader1990) [inline argus/util.py]: add the required unit test and all good

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/config/arestor.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2017 Cloudbase Solutions Srl
```
Comment (ader1990) [inline argus/config/arestor.py:1]: Here we need an extensive commit description on what is arestor and why we use arestor and what are the current limitations.

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/config/arestor.py [hunk 5]
```diff
@@ -37,6 +37,9 @@ def __init__(self, config):
             cfg.StrOpt(
                 "secret", default=None,
                 help="The secret for the arestor user"),
+            cfg.BoolOpt(
+                "delete_metadata", default=True,
+                help="Delete posted metadata when the scenaro is over."),
```
Comment (ader1990) [inline argus/config/arestor.py]: typo*

File: argus/backends/windows.py [hunk 6]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
                                        transport_protocol=protocol)
 
     remote_client = util.cached_property(get_remote_client, 'remote_client')
+
+
+class BackendMetadataProviderMixin(
+        base_metadata_provider.BaseMetadataProviderMixing):
+    """Backend Metadata Provider mixing."""
```
Comment (ader1990) [inline argus/backends/windows.py]: Use metadata consistently and not meta-data.

File: argus/backends/windows.py [hunk 7]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
                                        transport_protocol=protocol)
 
     remote_client = util.cached_property(get_remote_client, 'remote_client')
+
+
+class BackendMetadataProviderMixin(
+        base_metadata_provider.BaseMetadataProviderMixing):
```
Comment (ader1990) [inline argus/backends/windows.py]: BaseMetadataProviderMixin

File: argus/backends/windows.py [hunk 8]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
...
+    def get_url(self, service_type):
+        """Return the metadata url."""
+        return util.DETAULT_METADATA_URL.get(service_type)
+
+    def get_password(self):
+        """Get the encripted password."""
+        return self.instance_password()
+
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with post public keys."""
```
Comment (ader1990) [inline argus/backends/windows.py]: ssh public keys

File: argus/backends/windows.py [hunk 9]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
...
+class BackendMetadataProviderMixin(
+        base_metadata_provider.BaseMetadataProviderMixing):
+    """Backend Metadata Provider mixing."""
+
+    def get_url(self, service_type):
+        """Return the metadata url."""
+        return util.DETAULT_METADATA_URL.get(service_type)
+
+    def get_password(self):
+        """Get the encripted password."""
```
Comment (ader1990) [inline argus/backends/windows.py]: encrypted*

File: argus/backends/windows.py [hunk 10]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
...
+        return self.instance_password()
+
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with post public keys."""
+        return {0: self.get_ssh_pubkeys}
+
+    def set_ssh_pubkeys(self, public_key):
+        """Get a dictionary with post public keys."""
```
Comment (ader1990) [inline argus/backends/windows.py]: with ssh public keys*

File: argus/backends/windows.py [hunk 11]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
...
+        raise NotImplementedError()
+
+    def delete_all_data(self):
+        """Clean up all metadata."""
+        pass
+
+    def prepare_metadata(self, service_type):
+        """Clean up all metadata."""
```
Comment (ader1990) [inline argus/backends/windows.py]: Prepare or create the metadata

File: argus/metadata_provider/__init__.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+from argus.metadata_provider import arestor_provider
+from argus import util
+
+
+def _get_provider_map(recipe, backend):
+    """Get a map with all metadata providers."""
+    arestor_metadata_provider = arestor_provider.ArestorMetadataProvider(
+        backend=backend, recipe=recipe)
+    return {
```
Comment (ader1990) [inline argus/metadata_provider/__init__.py]: put a note here what backend means

File: argus/metadata_provider/arestor_provider.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+from argus import log as argus_log
+from argus import util
+from argus.metadata_provider import base
+
+
+LOG = argus_log.LOG
+CONFIG = argus_config.CONFIG
+
+
+class ArestorMetadataProviderMixin(base.BaseMetadataProviderMixing):
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: Mixin*

File: argus/metadata_provider/arestor_provider.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    """Mixin for Arestor Metadata Provider."""
+
+    # pylint: disable=no-member
+    def get_url(self, service_type):
+        """Return the metadata url."""
+        return self._arestor_client.get_url()
+
+    def get_password(self):
+        """Get the encripted password."""
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: ditto

File: argus/metadata_provider/arestor_provider.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        return self._arestor_client.get_password()
+
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with post public keys."""
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: ditto

File: argus/metadata_provider/arestor_provider.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+        return self._arestor_client.get_ssh_pubkeys()
+
+    def set_ssh_pubkeys(self, public_key):
+        """Get a dictionary with post public keys."""
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: ditto

File: argus/metadata_provider/arestor_provider.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+        :param backend: The Instance backend.
+        :param recipe: The recipe used.
+        """
+        super(ArestorMetadataProvider, self).__init__(backend, recipe)
+        self._arestor_client = arestor_client.ArestorClient(
+            base_url=CONFIG.arestor.base_url,
+            api_key=CONFIG.arestor.api_key,
+            secret=CONFIG.arestor.secret,
+            client_id="instance-" + backend.instance_server()['name'])
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: use string formatting

File: argus/metadata_provider/arestor_provider.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+        """Return the metadata namespace."""
+        # NOTE(mmicu): for Openstack we have the service_type set to http
+        return service_type if service_type != "http" else "openstack"
+
+    def setup_metadata(self, service_type):
+        """Set up the required metadata."""
+        super(ArestorMetadataProvider, self).setup_metadata(service_type)
+        namespace = self._get_namespace(service_type)
+        instance_server = self._backend.instance_server()
+        name = instance_server["name"]
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: use .get instead of []

File: argus/metadata_provider/arestor_provider.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+        self._arestor_client.set_user_data(self._backend.userdata)
+        self._arestor_client.set_metadata(json.dumps(self._backend.metadata))
+
+        self._arestor_client.set_project_id("project-{}".format(name))
+        self._arestor_client.set_launch_index("0")
+        self._arestor_client.set_availability_zone(
+            instance_server.get("OS-EXT-AZ:availability_zone",
+                                "az-{}".format(name)))
+        self._arestor_client.set_random_seed("random-seed-{}".format(name))
+        self._arestor_client.set_uuid(instance_server["id"])
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: ditto

File: argus/metadata_provider/base.py [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+from argus import log as argus_log
+from argus import util
+
+
+LOG = argus_log.LOG
+CONFIG = argus_config.CONFIG
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseMetadataProviderMixing(object):
```
Comment (ader1990) [inline argus/metadata_provider/base.py]: Mixin*

File: argus/metadata_provider/base.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+    """Base class for a MetadataProviderMixing."""
```
Comment (ader1990) [inline argus/metadata_provider/base.py]: ditto

File: argus/metadata_provider/base.py [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+
+    # pylint: disable=no-member
+    @abc.abstractmethod
+    def get_url(self, service_type):
+        """Return the metadata url."""
+        pass
+
+    @abc.abstractmethod
+    def get_password(self):
+        """Get the encripted password."""
```
Comment (ader1990) [inline argus/metadata_provider/base.py]: ditto

File: argus/metadata_provider/base.py [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+        pass
+
+    @abc.abstractmethod
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with post public keys."""
```
Comment (ader1990) [inline argus/metadata_provider/base.py]: ditto

File: argus/metadata_provider/base.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+        pass
+
+    @abc.abstractmethod
+    def set_ssh_pubkeys(self, public_key):
+        """Get a dictionary with post public keys."""
```
Comment (ader1990) [inline argus/metadata_provider/base.py]: ditto

File: argus/metadata_provider/base.py [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+        pass
+
+    @abc.abstractmethod
+    def delete_all_data(self):
+        """Clean up all metadata."""
+        pass
+
+    def prepare_metadata(self, service_type):
+        """Clean up all metadata."""
```
Comment (ader1990) [inline argus/metadata_provider/base.py]: ditto

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/backends/heat/heat_backend.py [hunk 26]
```diff
@@ -272,8 +272,7 @@ def reboot_instance(self):
     def instance_password(self):
```
Comment (ader1990) [inline argus/backends/heat/heat_backend.py]: encrypted*

**Review by ader1990 — CHANGES_REQUESTED** [review]

**Review by ader1990 — CHANGES_REQUESTED** [review]
Most of the issues are of naming, typos and quick fixable ones. The big picture seems good.

File: argus/util.py (hunk 3 above)
Comment (mateimicu) [reply argus/util.py]: Done

File: argus/backends/heat/heat_backend.py (hunk 26 above)
Comment (mateimicu) [reply argus/backends/heat/heat_backend.py]: Done

File: argus/backends/tempest/manager.py [hunk 27]
```diff
@@ -122,22 +107,17 @@ def reboot_instance(self, instance_id):
...
-            the password.
         """
         encoded_password = self.servers_client.show_password(
             instance_id)
-        with _create_tempfile(keypair.private_key) as tmp:
-            return util.decrypt_password(
-                private_key=tmp,
-                password=encoded_password['password'])
+
+        return encoded_password['password']
```
Comment (ader1990) [inline argus/backends/tempest/manager.py]: use .get

File: argus/backends/windows.py [hunk 28]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
                                        transport_protocol=protocol)
 
     remote_client = util.cached_property(get_remote_client, 'remote_client')
+
+
+class BaseMetadataProviderMixin(
+        base_metadata_provider.BaseMetadataProviderMixin):
+    """Backend Metadata Provider mixing."""
```
Comment (ader1990) [inline argus/backends/windows.py]: mixin*

File: argus/backends/windows.py [hunk 29]
```diff
@@ -37,3 +38,32 @@ def get_remote_client(self, username=None, password=None,
...
+    def get_password(self):
+        """Get the encrypted password."""
+        return self.instance_password()
+
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with ssh public keys."""
+        return {0: self.get_ssh_pubkeys}
+
+    def set_ssh_pubkeys(self, public_key):
+        """Get a dictionary with ssh public keys."""
```
Comment (ader1990) [inline argus/backends/windows.py]: Set*

File: argus/config/arestor.py [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+                help="The base URL for this service"),
+            cfg.StrOpt(
+                "api_key", default=None,
+                help="The api key for the arestor user."),
+            cfg.StrOpt(
+                "secret", default=None,
+                help="The secret for the arestor user"),
+            cfg.BoolOpt(
+                "delete_metadata", default=True,
+                help="Delete posted metadata when the scenario is over."),
```
Comment (ader1990) [inline argus/config/arestor.py]: is finished.

File: argus/metadata_provider/arestor_provider.py [hunk 31]
```diff
@@ -0,0 +1,116 @@
...
+    def get_password(self):
+        """Get the encrypted password."""
+        return self._arestor_client.get_password()
+
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with the ssh public keys."""
+        return self._arestor_client.get_ssh_pubkeys()
+
+    def set_ssh_pubkeys(self, public_key):
+        """Get a dictionary with the ssh public keys."""
```
Comment (ader1990) [inline argus/metadata_provider/arestor_provider.py]: ditto

File: argus/metadata_provider/base.py [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+        pass
+
+    @abc.abstractmethod
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with ssh hpublic keys."""
+        pass
+
+    @abc.abstractmethod
+    def set_ssh_pubkeys(self, public_key):
+        """Get a dictionary with ssh public keys."""
```
Comment (ader1990) [inline argus/metadata_provider/base.py]: ditto

File: argus/recipes/cloud/base.py [hunk 33]
```diff
@@ -136,6 +136,11 @@ def replace_code(self):
         """Do whatever is necessary to replace the code for Cloudbase-Init."""
 
     @abc.abstractmethod
+    def create_mock_metadata(self, service_type):
+        """Create the mocked meta-data."""
```
Comment (ader1990) [inline argus/recipes/cloud/base.py]: metadata*

File: argus/recipes/cloud/base.py [hunk 34]
```diff
@@ -160,6 +165,11 @@ def get_cb_init_confs(self):
         """Get the Cloudbase-Init configs from the instance."""
         pass
 
+    @abc.abstractmethod
+    def delete_mock_metadata(self):
+        """Delete the mocked meta-data."""
```
Comment (ader1990) [inline argus/recipes/cloud/base.py]: ditto

File: argus/recipes/cloud/windows.py [hunk 35]
```diff
@@ -260,14 +265,30 @@ def wait_cbinit_finalization(self):
...
         self._backend.remote_client.manager.wait_cbinit_service()
 
+    @staticmethod
+    def _get_namespace(service_type):
+        """Return the metadata namespace."""
+        # NOTE(mmicu): for Openstack we have the service_type set to http
+        return service_type if service_type != "http" else "openstack"
+
+    def create_mock_metadata(self, service_type):
+        """Create the mocked meta-data."""
```
Comment (ader1990) [inline argus/recipes/cloud/windows.py]: ditto

File: argus/recipes/cloud/windows.py [hunk 36]
```diff
@@ -260,14 +265,30 @@ def wait_cbinit_finalization(self):
...
+        self.metadata_provider = metadata_provider.get_provider(
+            self, self._backend, service_type)
+
+        self.metadata_provider.prepare_metadata(service_type)
+
+    def delete_mock_metadata(self):
+        """Delete the mocked meta-data."""
```
Comment (ader1990) [inline argus/recipes/cloud/windows.py]: ditto

**Review by ader1990 — CHANGES_REQUESTED** [review]

File: argus/metadata_provider/__init__.py (hunk 12 above)
Comment (mateimicu) [reply argus/metadata_provider/__init__.py]: Done

File: argus/metadata_provider/base.py [hunk 37]
```diff
@@ -0,0 +1,85 @@
...
+        pass
+
+    @abc.abstractmethod
+    def get_password(self):
+        """Get the encrypted password."""
+        pass
+
+    @abc.abstractmethod
+    def get_ssh_pubkeys(self):
+        """Get a dictionary with ssh hpublic keys."""
```
Comment (paulaCrismaru) [inline argus/metadata_provider/base.py]: public*

File: argus/config/arestor.py [hunk 38]
```diff
@@ -0,0 +1,50 @@
...
+
+    """Config options available for the Arestor metadata service."""
+
+    def __init__(self, config):
+        super(ArestorOptions, self).__init__(config,
+                                             group="arestor")
+        self._options = [
+            cfg.StrOpt(
+                "base_url",
+                default="http://0.0.0.0:8080",
```
Comment (ader1990) [inline argus/config/arestor.py]: default should be localhost: 127.0.0.1 -- you cannot make a request to 0/0, you can only listen at 0/0

**Review by ader1990 — CHANGES_REQUESTED** [review]

**Review by ader1990 — APPROVED** [review]

**Pull request merged**